src/irmin-pack/unix/sparse_file.ml

```diff
@@ -324,15 +172,78 @@ module Make (Io : Io.S) = struct
     let len = min max_len max_entry_len in
     Io.read_exn t.data ~off:poff ~len buf
 
-  let write_exn t ~off ~len str =
-    let poff, max_entry_len = get_poff t ~off in
-    assert (len <= max_entry_len);
-    Io.write_exn t.data ~off:poff ~len str
-
   let next_valid_offset { mapping; _ } ~off =
     match Mapping_file.find_nearest_geq mapping off with
     | None -> None
     | Some entry ->
         let open Int63.Syntax in
         Some (if entry.off < off then off else entry.off)
+
+  module Wo = struct
+    type nonrec t = t
+
+    let open_wo ~mapping ~data = open_ ~readonly:false ~mapping ~data
+
+    let write_exn t ~off ~len str =
+      let poff, max_entry_len = get_poff t ~off in
+      assert (len <= max_entry_len);
+      Io.write_exn t.data ~off:poff ~len str
+
+    let fsync t = Io.fsync t.data
+    let close = close
+  end
+
+  module Ao = struct
+    module Ao = Append_only_file.Make (Io) (Errs)
+
+    type t = { mapping : Ao.t; data : Ao.t }
+
+    let create ~mapping ~data =
+      let open Result_syntax in
+      let ao_create path =
+        Ao.create_rw ~path ~overwrite:false ~auto_flush_threshold:1_000_000
+          ~auto_flush_procedure:`Internal
+      in
+      let* mapping = ao_create mapping in
+      let+ data = ao_create data in
+      { mapping; data }
+
+    let open_ao ~mapping ~data =
+      let open Result_syntax in
+      let ao_open path =
+        (* TODO: the expected [end_poff] must be obtained from the control file *)
+        let* end_poff = Io.size_of_path path in
+        Ao.open_rw ~path ~end_poff ~dead_header_size:0
+          ~auto_flush_threshold:1_000_000 ~auto_flush_procedure:`Internal
+      in
+      let* mapping = ao_open mapping in
+      let+ data = ao_open data in
+      { mapping; data }
+
+    let append_seq_exn t ~off seq =
+      (* TODO: check that [off] is larger or equal to the previous entry ending offset. *)
```

Comment:
Reminder to capture this in the miscellaneous section of our überissue.

Reply:
I've added a sanity check (that should not raise if the future GC code is correct)